pkg/finality-grandpa/bridge_state.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"sync"
+)
+
+type waker struct {
+	mtx    sync.RWMutex
+	wakeCh chan any
+}
+
+func newWaker() *waker {
+	return &waker{wakeCh: make(chan any, 1000)}
+}
+
+func (w *waker) wake() {
+	w.mtx.RLock()
+	defer w.mtx.RUnlock()
+	if w.wakeCh == nil {
+		return
+	}
+	go func() {
+		select {
+		case w.wakeCh <- nil:
+		default:
+		}
+	}()
+}
+
+func (w *waker) channel() chan any {
+	return w.wakeCh
+}
+
+func (w *waker) register(waker *waker) {
+	w.mtx.Lock()
+	defer w.mtx.Unlock()
+	w.wakeCh = waker.wakeCh
+}
+
+// round state bridged across rounds.
+type bridged[Hash, Number any] struct {
+	inner RoundState[Hash, Number]
+	// registered map[chan State[Hash, Number]]any
+	waker *waker
+	sync.RWMutex
+}
+
+func (b *bridged[H, N]) update(new RoundState[H, N]) {
+	b.Lock()
+	b.inner = new
+	b.waker.wake()
+	b.Unlock()
+}
+
+func (b *bridged[H, N]) get(waker *waker) RoundState[H, N] {
+	b.RLock()
+	defer b.RUnlock()
+	b.waker.register(waker)
+	return b.inner
+}
+
+// A prior view of a round-state.
+type priorView[Hash, Number any] struct {
+	bridged *bridged[Hash, Number]
+}
```

Comment:
why do we need to structs that holds the same value? can we turn it in a single struct and then use interface to constraint the behavior:

```go

type bridgeVIew[H, N any] struct {
    bridged *bridged[Hash, Number]
}

// Push an update to the latter view.
func (pv *bridgeVIew[H, N]) update(new RoundState[H, N]) { //skipcq: RVV-B0001
	pv.bridged.update(new)
}

// Fetch a handle to the last round-state.
func (lv *bridgeVIew[H, N]) get(waker *waker) (state RoundState[H, N]) { //skipcq: RVV-B0001
	return lv.bridged.get(waker)
}

type LatterView[H, N any]  interface {
    get(waker *waker) (state RoundState[H, N])
}

type PriorView[H, N any]  interface {
    update(new RoundState[H, N])
}

func bridgeState[Hash, Number any](initial RoundState[Hash, Number]) (
	LatterView[Hash, Number],
	PriorView[Hash, Number],
) {
	br := bridged[Hash, Number]{
		inner: initial,
		waker: newWaker(),
	}
	return bridgeVIew[Hash, Number]{&br}, bridgeVIew[Hash, Number]{&br}
}
```

Reply:
Thanks for this! I ended having to change a bunch of the attributes that store these types.